File: data-structure/queue/circular_queue.py
MAX_QSIZE = 10


class CircularQueue:
    def __init__(self):
        self.front = 0
        self.end = 0
        self.items = [None] * MAX_QSIZE

    def is_empty(self):
        return self.front == self.end

    def is_full(self):
        # rear 는 항상 front 보다 앞에 있거나 같다
        # index 는 원형이라 더 작을 수 있지만 rear 를 넘어서 front 가 갈 수는 없다.
        return self.front == (self.end + 1) % MAX_QSIZE

    def clear(self):
        return self.front == self.end

    def length(self):
        return (self.end - self.front + MAX_QSIZE) % MAX_QSIZE

    def enqueue(self, item):
        if not self.is_full():
            self.end = (self.end + 1) % MAX_QSIZE
            self.items[self.end] = item

    def dequeue(self):
        if not self.is_empty():
            self.front = (self.front + 1) % MAX_QSIZE
            return self.items[self.front]

    def print_queue(self):
        arr = []
        if self.front < self.end:
            arr = self.items[self.front + 1: self.end + 1]
        elif self.front == self.end:
            arr = []
        else:
            arr = self.items[:self.end + 1] + self.items[self.front + 1:]
        print("#{} {} {}".format(self.front, self.end, arr))


cq = CircularQueue()
cq.enqueue(1)
cq.dequeue()
cq.print_queue()
cq.enqueue(2)
cq.dequeue()
cq.print_queue()
cq.enqueue(3)
cq.enqueue(4)
cq.enqueue(5)
cq.print_queue()

File: data-structure/linked-list/CircularQueue.java
package com.example;

public class CircularQueue {

    private Node head;
    private Node tail;
    private final int maxSize;
    private int size = 0;

    public CircularQueue(int k) {
        maxSize = k;
    }

    public boolean enQueue(int value) {
        if (isFull()) {
            return false;
        }

        if (isEmpty()) {
            head = new Node(value);
            tail = head;
        } else {
            // 기존에 tail 이 참조하던 head 에 대한 참조 제거
            Node oldTail = tail;
            Node newTail = new Node(value);
            oldTail.next = newTail;
            tail = newTail;
            tail.next = head;
        }

        size += 1;
        return true;
    }

    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        }

        if (size == 1) {
            head = null;
            tail = null;
        } else {
            Node newHead = head.next;
            head = newHead;
            tail.next = newHead;
        }

        size -= 1;
        return true;
    }

    public int Front() {
        if (isEmpty()) {
            return -1;
        }
        return head.value;
    }

    public int Rear() {
        if (isEmpty()) {
            return -1;
        }
        return tail.value;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return maxSize == size;
    }

    static class Node {
        int value;
        Node next;

        public Node(int value) {
            this.value = value;
        }
    }
}
